daemon/containerd/image.go

```diff
@@ -97,7 +126,9 @@ func (i *ImageService) presentImages(ctx context.Context, img containerdimages.I
 		return nil, err
 	}
 	if len(presentImages) == 0 {
-		return nil, errdefs.NotFound(errors.New("no matching image found"))
+		return nil, &errPlatformNotFound{
+			imageRef: img.Name,
+		}
```

Comment:
Is this situation is only if the manifest index does not provide a usable platform, or also if it does (but the content isn't present)?

Reply:
Both for now.

Reply:
To add to the above; the reason I was wondering is because I know we have logic on the client-side where we have special handling for `not found` errors; when doing a `docker run` (or `docker create`), and we get a 404, the CLI initiates a `docker pull` (image not found; let's try pulling it).

Very likely not for this PR, but something we need to start looking at, because with multi-arch there will be more permutations of that;

- The image is missing as a whole; we'll try pulling the image (manifest-index) and the desired platform.
- We pulled the manifest-index, and it has an image for the given platform, but it's not yet pulled. In this case, we can pull the missing platform (but keep the manifest-index at the same version)
- We pulled the manifest-index, but it does not contain the given platform

The last one is the tricky one, at least with the current logic not being granular enough;

- We can either bail out (we have the manifest-index, but unfortunately it doesn't have this platform)
- We can try pulling the manifest-index to see if perhaps a current version of it _does_ have the platform (maybe the image was updated since we pulled the manifest-index?)

Either of the above are "valid", but we don't have a lot of granularity for this currently. The last one also can be confusing, as pulling the manifest-index may mean that any image (for other platform) may now become "dangling" as they no longer match the digest of the newly pulled manifest-index.